pkg/local-gadget-manager/local-gadget-manager.go

```diff
@@ -0,0 +1,409 @@
+		}
+		tracerNamespacedName := tracer.traceResource.ObjectMeta.Namespace +
+			"/" + tracer.traceResource.ObjectMeta.Name
+		gadgetOperation.Operation(tracerNamespacedName, tracer.traceResource)
+	}
+
+	fmt.Printf("State: %s\n", tracer.traceResource.Status.State)
+	fmt.Printf("OperationError: %s\n", tracer.traceResource.Status.OperationError)
+	fmt.Printf("Output: %s\n", tracer.traceResource.Status.Output)
+
+	return nil
+}
+
+func (l *LocalGadgetManager) Show(name string) error {
+	return l.Operation(name, "")
+}
+
+func (l *LocalGadgetManager) Delete(name string) error {
+	tracer, ok := l.tracers[name]
+	if !ok {
+		return fmt.Errorf("unknown trace")
+	}
+
+	factory, ok := l.traceFactories[tracer.gadget]
+	if !ok {
+		return fmt.Errorf("unknown gadget")
+	}
+	factory.Delete("gadget/" + name)
+	delete(l.tracers, name)
+	return nil
+}
+
+func (l *LocalGadgetManager) PublishEvent(tracerID string, line string) error {
+	name := strings.TrimPrefix(tracerID, "trace_gadget_")
+	t, ok := l.tracers[name]
+	if !ok {
+		return fmt.Errorf("cannot find tracer: unknown tracer %q", tracerID)
+	}
+
+	t.gadgetStream.Publish(line)
+	return nil
+}
+
+func (l *LocalGadgetManager) Stream(name string, stop chan struct{}) (chan string, error) {
+	t, ok := l.tracers[name]
+	if !ok {
+		return nil, fmt.Errorf("cannot find trace %q", name)
+	}
+
+	out := make(chan string)
+
+	ch := t.gadgetStream.Subscribe()
+
+	go func() {
+		if stop == nil {
+			for len(ch) > 0 {
+				line := <-ch
+				out <- line.Line
+			}
+			t.gadgetStream.Unsubscribe(ch)
+			close(out)
+		} else {
+			for {
+				select {
+				case <-stop:
+					t.gadgetStream.Unsubscribe(ch)
+					close(out)
+					return
+				case line := <-ch:
+					out <- line.Line
+				}
+			}
+		}
+	}()
+	return out, nil
+}
+
+func (l *LocalGadgetManager) LookupMntnsByContainer(namespace, pod, container string) uint64 {
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		if container != c.Name {
+			continue
+		}
+		return c.Mntns
+	}
+	return 0
+}
+
+func (l *LocalGadgetManager) LookupMntnsByPod(namespace, pod string) map[string]uint64 {
+	ret := make(map[string]uint64)
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		ret[c.Name] = c.Mntns
+	}
+	return ret
+}
+
+func (l *LocalGadgetManager) LookupPIDByContainer(namespace, pod, container string) uint32 {
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		if container != c.Name {
+			continue
+		}
+		return c.Pid
+	}
+	return 0
+}
+
+func (l *LocalGadgetManager) LookupPIDByPod(namespace, pod string) map[string]uint32 {
+	ret := make(map[string]uint32)
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		ret[c.Name] = c.Pid
+	}
+	return ret
+}
```

Comment:
I think all of these are duplicated in gadget tracer manager. Is it too difficult to abstract the code in a way that we don't have to duplicate them?

Reply:
Makes sense. I'll work on this refactoring first.

Reply:
Done in https://github.com/kinvolk/inspektor-gadget/pull/298